crates/tx5/src/peer.rs

```diff
@@ -101,41 +97,26 @@ impl Peer {
     }
 
     /// This is initially false, but can transition into true.
+    ///
     /// If establishing a webrtc connection fails, it will remain false.
     pub fn is_using_webrtc(&self) -> bool {
-        if let MaybeReady::Ready(r) = &*self.ready.lock().unwrap() {
-            r.is_using_webrtc()
-        } else {
-            false
-        }
+        self.ready
+            .query_ready(|c| c.is_using_webrtc())
+            .unwrap_or_default()
     }
 
     /// Get connection statistics.
     pub fn get_stats(&self) -> ConnStats {
-        if let MaybeReady::Ready(r) = &*self.ready.lock().unwrap() {
-            r.get_stats()
-        } else {
-            ConnStats::default()
-        }
-    }
-
-    /// This future resolves when the connection is ready to use.
-    pub async fn ready(&self) {
-        let w = match &*self.ready.lock().unwrap() {
-            MaybeReady::Ready(_) => return,
-            MaybeReady::Wait(w) => w.clone(),
-        };
-
-        let _ = w.acquire().await;
+        self.ready
+            .query_ready(|c| c.get_stats())
+            .unwrap_or_default()
     }
 
-    /// Send data to the remote peer over this connection.
-    pub async fn send(&self, msg: Vec<u8>) -> Result<()> {
-        let conn = match &*self.ready.lock().unwrap() {
-            MaybeReady::Ready(c) => c.clone(),
-            _ => return Err(Error::other("not ready")),
-        };
-        conn.send(msg).await
+    /// This future resolves when the connection is ready to use or has failed to connect.
+    ///
+    /// If the connection is not usable, it will return `None`.
+    pub async fn wait_for_ready(&self) -> Option<DynBackCon> {
```

Comment:
Note the change here. I'm slightly breaking encapsulation by handing out the connection but it is only handed out within the module. It's given back to the endpoint to use.

By having the "ready" operation return the connection when the connection is available we can just have one function and let the caller do their own send.